src/util/ShapeEditor.js

```diff
@@ -0,0 +1,1570 @@
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+/**
+ * @exports ShapeEditor
+ */
+define([
+        '../geom/Angle',
+        '../shapes/Annotation',
+        '../shapes/AnnotationAttributes',
+        '../error/ArgumentError',
+        '../util/Color',
+        '../util/Font',
+        '../util/Insets',
+        '../geom/Location',
+        '../util/Logger',
+        '../shapes/Path',
+        '../shapes/Placemark',
+        '../shapes/PlacemarkAttributes',
+        '../geom/Position',
+        '../layer/RenderableLayer',
+        '../shapes/ShapeAttributes',
+        '../shapes/SurfaceEllipse',
+        '../shapes/SurfaceCircle',
+        '../shapes/SurfacePolygon',
+        '../shapes/SurfacePolyline',
+        '../shapes/SurfaceRectangle',
+        '../shapes/SurfaceShape',
+        '../geom/Vec2',
+        '../geom/Vec3'
+    ],
+    function (Angle,
+              Annotation,
+              AnnotationAttributes,
+              ArgumentError,
+              Color,
+              Font,
+              Insets,
+              Location,
+              Logger,
+              Path,
+              Placemark,
+              PlacemarkAttributes,
+              Position,
+              RenderableLayer,
+              ShapeAttributes,
+              SurfaceEllipse,
+              SurfaceCircle,
+              SurfacePolygon,
+              SurfacePolyline,
+              SurfaceRectangle,
+              SurfaceShape,
+              Vec2,
+              Vec3) {
+        "use strict";
+
+        /**
+         * @alias ShapeEditor
+         * @constructor
+         * @classdesc Provides a user interface for editing a shape and performs editing. Depending on the shape type,
+         * the shape is shown with control points for vertex locations and size. All shapes are shown with a handle that
+         * provides rotation.
+         * <p/>
+         * Drag on the shape's body moves the whole shape. Drag on a control point performs the action
+         * associated with that control point. The editor provides vertex insertion and removal for SurfacePolygon and
+         * SurfacePolyline. Shift-click when the cursor is over the shape inserts a control
+         * point at the cursor's position. Alt-click when the cursor is over a control point removes that control point.
+         * <p/>
+         * This editor supports all surface shapes except SurfaceImage.
+         * @param {WorldWindow} worldWindow The World Window to associate this shape editor controller with.
+         * @throws {ArgumentError} If the specified world window is null or undefined.
+         */
+        var ShapeEditor = function (worldWindow, shape) {
```

Comment:
@pdavidc Should we call this class SurfaceShapeEditor for clarify?

Reply:
Good question @AkeluX. We'll likely need editors for other shape types, like 3D polygons and the 3D airspace shapes we'll eventually develop. Do you think that this shape editor would be extended to support those shapes?
(Keep in mind, 3D shapes are displayed as surface shapes in 2D projection mode)